### PR TITLE
Add Tags Feature Branch

### PR DESCRIPTION
**Pull request by cyndichin**
## Summary
Users can add tags to a saved item by using existing tags or inputing new tags via text field

## References 
IN-227, IN-231, IN-232

## Implementation Details
See previous PRs for implementation details:
* https://github.com/Pocket/pocket-ios/pull/223
* https://github.com/Pocket/pocket-ios/pull/227
* https://github.com/Pocket/pocket-ios/pull/232

## Test Steps
- [ ] Given a user has navigated to the Reader View for a Saved Item,
and the user taps on the overflow menu button on the top corner, 
then the user should see an Add Tags option.

- [x] Given a user has navigated to their Saves,
and the user taps on the overflow menu button, 
then the user should see an Add Tags option.

- [x] Given a user has navigated to their Archive list,
and the user taps on the overflow menu button, 
then the user should see an Add Tags option.

- [x] Given a user has tapped on the Add Tags option for a Saved Item,
and the user views the text field (which is attached to the keyboard),
then the user should be able to enter and add tags.

- [x] Given a user has tapped on the Add Tags option for a Saved Item,
and the user inputs a tag into the text field, 
then the user should be able save it.

- [x] Given a user is on the Add Tags screen, 
and the user views the all tags section, 
then the user should be able to see existing tags (unless included in the text field).

- [x] Given a user is on the Add Tags screen, 
and the user taps on one of the existing tags, 
then the user should see it appear in the text field.

- [x] Given a user has tapped on the Add Tags option for a Saved Item,
and the user taps on a tag, 
then the user should be able to delete it. 

Note: Changes should also reflect server side and reflect on web.

## PR Checklist:
- [x] Added Unit / UI tests
- [x] Self Review (review, clean up, documentation, run tests)
- [x] Basic Self QA

## Screenshots
https://user-images.githubusercontent.com/6743397/187524957-5751758d-e932-4fb8-be8b-4a283ce59c74.mp4

### CODE REVIEW

File: PocketKit/Sources/PocketKit/Home/CompactHomeCoordinator.swift [hunk 1]
```diff
@@ -252,6 +257,13 @@ class CompactHomeCoordinator: NSObject {
         guard !isResetting, let alert = alert else { return }
         viewController.present(UIAlertController(alert), animated: !isResetting)
     }
+    
+    func present(_ viewModel: AddTagsViewModel?) {
+        guard let viewModel = viewModel else { return }
```
Comment (jacobsimeon) [inline PocketKit/Sources/PocketKit/Home/CompactHomeCoordinator.swift:262]: When presenting modal view controllers, we need to account for the case where a user is changing from compact to regular. Since modal view controllers persist during the transition, we want to "short circuit" this logic so we don't present a second time. 

TLDR; we need to `guard !isResetting else { return }` here 😆

File: PocketKit/Sources/PocketKit/Main/RegularMainCoordinator.swift [hunk 2]
```diff
@@ -258,6 +263,13 @@ extension RegularMainCoordinator: ModalContentPresenting {
         safariVC.delegate = self
         splitController.present(safariVC, animated: !isResetting)
     }
+    
+    func present(_ viewModel: AddTagsViewModel?) {
+        guard let viewModel = viewModel else { return }
```
Comment (jacobsimeon) [inline PocketKit/Sources/PocketKit/Main/RegularMainCoordinator.swift:268]: ditto here, `guard !isResetting else { return }`

File: PocketKit/Sources/Sync/RemoteMapping/SavedItem+RemoteMapping.swift [hunk 3]
```diff
@@ -34,7 +34,18 @@ extension SavedItem {
...
-
+        
+        if let tags = tags {
+            removeFromTags(tags)
+        }
+        
+        remote.tags?.compactMap { $0 }.forEach({ remoteTag in
+            let tag = space.fetchOrCreateTag(byName: remoteTag.name)
+            addToTags(tag)
+        })
```
Comment (jacobsimeon) [inline PocketKit/Sources/Sync/RemoteMapping/SavedItem+RemoteMapping.swift:45]: Could simplify this to something like:

```swift
tags = remote.tags?.compactMap { $0 }.map { remoteTag in
    space.fetchOrCreateTag(byName: remoteTag.name)
}
```

**Review by jacobsimeon — CHANGES_REQUESTED** [review]
Looks great, just need to guard against presenting a duplicate modal view controller in a couple cases.

File: PocketKit/Sources/PocketKit/Home/CompactHomeCoordinator.swift (hunk 1 above)
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/Home/CompactHomeCoordinator.swift]: oh thanks! I tried to replicate this issue but wasn't sure how to, but I updated the change. We can discussed next week on how to reproduce the issue.

File: PocketKit/Sources/Sync/RemoteMapping/SavedItem+RemoteMapping.swift (hunk 3 above)
Comment (cyndichin) [reply PocketKit/Sources/Sync/RemoteMapping/SavedItem+RemoteMapping.swift]: updated!

**Review by jacobsimeon — APPROVED** [review]

**Comment by audie1** [comment]
Issues found during testing are listed here:
https://pocket.slack.com/archives/C4DL4QSU8/p1662491374184089

Passing the PR based on the core functionality working 💯 good work @cyndichin
